### PR TITLE
Move control point block removal to hwt

### PR DESCRIPTION
**Pull request by ltratt**
This PR moves the "ignore the first block" logic, which is only relevant for hwt, into hwt. The first commit does this brutally but makes what's going on clear; the second commit does so more elegantly and does a partial check that we've removed the right thing.

### CODE REVIEW

File: yktracec/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -1052,9 +1052,7 @@ class JITModBuilder {
   Module *createModule() {
     size_t CurBBIdx;
     size_t CurInstrIdx;
-    // Iterate over the blocks of the trace. The first block contains the
```
Comment (vext01) [inline yktracec/src/jitmodbuilder.cc:1055]: Shall we add a comment saying what the first block should be now?

You might be able to add an assertion that the first block is a successor of the block containing the control point?

Comment (ltratt) [reply yktracec/src/jitmodbuilder.cc:1055]: TBH, I don't know what the first block should be, nor do I know how to encode that assertion!

Comment (vext01) [reply yktracec/src/jitmodbuilder.cc:1055]: > I don't know what the first block should be

A successor of the block containing the control point call.

>  nor do I know how to encode that assertion

OK, not worth worrying about.

Comment (ltratt) [reply yktracec/src/jitmodbuilder.cc:1055]: Is it always a direct successor. I guess I assumed it could be a function call? Anyway, maybe not a big worry?

Comment (vext01) [reply yktracec/src/jitmodbuilder.cc:1055]: You're right, it could be an unmappable block inside a function call.

**Comment by vext01** [comment]
Hah, the new trace compiler backend looks at the block you just removed to find the trace inputs.

This answers a question in my PR:

https://github.com/ykjit/yk/pull/956/files#diff-0019b0cfa54cfb8c953a80e8ab4e0ce900b5fada3100f573c59c09feadd32fc6R172-R174

**Comment by ltratt** [comment]
Oh! What should we do?

**Comment by vext01** [comment]
If you want to keep this design, quickish fix: search the AOT IR for the control point call.

Better fix: extend ykllvm to encode where to find the control point call into the AOT IR. i.e. at compile time.

**Comment by ptersilie** [comment]
This change doesn't affect the way `jitmodbuilder.cc` currently searches for the control point call (see https://github.com/ykjit/yk/blob/master/yktracec/src/jitmodbuilder.cc#L933). Of course this won't work if there's multiple control point calls (note the `FIXME`), but its probably not worth fixing here.

We should merge this and then do the correct thing in the new trace builder, i.e. search for the control point block manually, which should be the predecessor of the first block in the trace.

**Comment by ptersilie** [comment]
@ltratt Do you want to squash?

**Comment by ltratt** [comment]
I think the two commits are OK separately.

**Comment by ltratt** [comment]
@ptersilie Now I remember the problem: I don't think the new code gen works with this change!

But that said, the error is a weird PT error I've never seen before (this is one for @vext01 to look at):

```
12:28:44    |jit-state: start-tracing
12:28:44 >> |thread '<unnamed>' panicked at hwtracer/src/pt/ykpt/mod.rs:453:13:
12:28:44    |assertion failed: taken
12:28:44    |note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
12:28:44    |fatal runtime error: failed to initiate panic, error 5
```

So maybe @ptersilie try rerunning CI and see if we get to the "genuine" error I think this PR might cause?

**Comment by vext01** [comment]
That's an error I've only ever seen when compressed returns are enabled. We disabled them a while back. When did you get this error? Today?

**Comment by vext01** [comment]
Right, in this PR. Investigating...

**Comment by vext01** [comment]
I re-ran the build and got the failure I'd expect (not the PT 'taken' error):
https://ci.soft-dev.org/#/builders/1/builds/3560/steps/2/logs/stdio

It is very curious that the "taken" thing comes up with compressed returns disabled. This suggests a PT packet parsing bug elsewhere.

I few month ago I spent 3 days scrutinising the compressed return logic and matching it up with the Intel manual. I was unable to find anything "dodgy".

**Comment by vext01** [comment]
I suggest we don't worry about the "taken" bug for now. It's very very rare, and we have an issue up for it, which I've just updated with our new information.

**Comment by ltratt** [comment]
@vext01 @ptersilie I suggest one of you try merging again (I expect the PR to fail but not for the PT reasons).

**Comment by vext01** [comment]
I linked to the error I expect above. We'd need to teach the new trace builder to find the control point from the successor block.

**Pull request merged**